astropy/coordinates/baseframe.py

```diff
@@ -424,39 +424,121 @@ def __init__(self, *args, copy=True, representation_type=None,
         representation_type = kwargs.pop('representation_type', representation_type)
 
         if representation_type is not None or differential_type is not None:
```

Comment:
Maybe also move this class to the new `_infer_representation_differential_type`?

Reply:
My initial philosophy was: "let these new `_infer` methods _not_ set any attributes, and do that only in `__init__`". That way, even though I distribute the logic and make `__init__` easier to read, the "mutation" only happens in one place. What do you think?

Reply:
Ah, I see I was not clear. I definitely agree that we should set the attributes here, I just thought we could deal with the `None, None` input also in your new method (which would then just return `self.default_representation, self.default_differential`), i.e., here you would simply have
```
        representation_type, differential_type = self._infer_representation_differential_type(
            representation_type, differential_type)
        self.set_representation_cls(representation_type, **differential_type)
```

Reply:
Actually, if your method returns `representation_type` also in the dict as the 'base' entry, you can pass it directly on (with name change):
```
self.set_representation_cls(**self._infer_representation_cls(
            representation_type, differential_type))
```

Reply:
Sorry, now see your next comment: I think I would just do
```
self._representation = self._infer_representation(representation_type, differential_type)
```
and let `_infer_representation` return a properly organized `dict`.

Reply:
Excellent suggestion, now the code is much more compact :heart_eyes: